### PR TITLE
Feature/implement memo

### PR DESCRIPTION
**Pull request by ZilongRRR**
# Description

Implement the memo in tx

* Send
image

* Activity 
image

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Send
* Memo to hex
https://github.com/hashgreen/hoogii-wallet/blob/816db3ca5c8a28258ba121f912c4ff49ad8c03c8/src/utils/CAT.ts#L105-L108

* Memo to byte
https://github.com/hashgreen/hoogii-wallet/blob/816db3ca5c8a28258ba121f912c4ff49ad8c03c8/src/utils/CAT.ts#L109

* Memo to program
https://github.com/hashgreen/hoogii-wallet/blob/816db3ca5c8a28258ba121f912c4ff49ad8c03c8/src/utils/CAT.ts#L137-L145

### Activity
* FIlter memo
https://github.com/hashgreen/hoogii-wallet/blob/feb57e258b64c55e81f9e7a32ad998a423158067/src/components/Transaction/Transaction.tsx#L60-L64

* Decode memo
https://github.com/hashgreen/hoogii-wallet/blob/feb57e258b64c55e81f9e7a32ad998a423158067/src/components/Transaction/Transaction.tsx#L66-L74

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* Firmware version:
* Hardware:
* Toolchain:
* SDK:

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# Todo
* Update the memo UI after new UI
* Backend would update the memos data, so the decode would need to be refactored
* Note that now the decode would not support "Chinese"

### CODE REVIEW

File: src/components/Transaction/Transaction.tsx [hunk 1]
```diff
@@ -197,17 +215,31 @@ const Transaction = ({
...
+                    <span className="capitalize">{t('transaction-memo')}</span>
+                    <div className="mt-1 text-tertiary">
+                        {decodedMemos?.map(
+                            (memo, index) =>
+                                memo && (
+                                    <span
+                                        className="break-words"
+                                        key={`${index}-${memo}`}
+                                    >
+                                        {`${memo.toString()} `}
```
Comment (hryjosn) [inline src/components/Transaction/Transaction.tsx]: @ZilongRRR  Do we need toString here?

File: src/utils/CAT.ts [hunk 2]
```diff
@@ -100,14 +101,26 @@ export class CAT extends Program {
         const puzzlehash = Program.fromBytes(
             addressInfo(targetAddress).hash
         ).toHex()
+
+        const memoHex = memo
+            .split('')
+            .map((c) => c.charCodeAt(0).toString(16).padStart(2, '0'))
+            .join('')
+        const memos: Uint8Array[] = memo.length === 0 ? [] : [fromHex(memoHex)]
```
Comment (hryjosn) [inline src/utils/CAT.ts:109]: Use fromBytes function from bls-signature library is better

File: src/components/Transaction/Transaction.tsx [hunk 3]
```diff
@@ -55,6 +56,23 @@ const Transaction = ({
...
         txType === ITxType.TX_TYPE_CAT_TRANSFER ||
         txType === ITxType.TX_TYPE_STANDARD_TRANSFER
+
+    // NOTE : if the tx is cat transfer, then the first memo is puzzlehash, so do not show it
+    const filteredMemo =
+        memos && txType === ITxType.TX_TYPE_CAT_TRANSFER
+            ? memos.slice(1)
+            : memos
+
+    const decodedMemos = filteredMemo.map((memo) =>
```
Comment (hryjosn) [inline src/components/Transaction/Transaction.tsx]: Using the function from bls-signature library to decodeMemo is better

File: src/components/Transaction/Transaction.tsx (hunk 1 above)
Comment (ZilongRRR) [reply src/components/Transaction/Transaction.tsx]: remove in [6b04294](https://github.com/hashgreen/hoogii-wallet/pull/12/commits/6b04294cfd0d938902a0136911fff2add794bcf2)

File: src/components/Transaction/Transaction.tsx (hunk 3 above)
Comment (ZilongRRR) [reply src/components/Transaction/Transaction.tsx]: now remove the decode function, the memo from backend would be decoded first

File: src/components/Transaction/Transaction.tsx [hunk 4]
```diff
@@ -197,17 +206,31 @@ const Transaction = ({
...
 
                 <div className="mt-1 text-tertiary" title={txId}>
                     {shortenHash(txId)}
                 </div>
+                <div className="pt-3 text-caption">
+                    <span className="capitalize">{t('transaction-memo')}</span>
+                    <div className="mt-1 text-tertiary">
+                        {filteredMemo?.map(
+                            (memo, index) =>
+                                memo && (
```
Comment (hryjosn) [inline src/components/Transaction/Transaction.tsx]: @ZilongRRR I think your can remove `memo &&` condition

**Review by hryjosn — CHANGES_REQUESTED** [review]
Except for the comment I left, others seem fine to me.

File: src/components/Transaction/Transaction.tsx (hunk 4 above)
Comment (ZilongRRR) [reply src/components/Transaction/Transaction.tsx]: [7858e6c](https://github.com/hashgreen/hoogii-wallet/pull/12/commits/7858e6c613a51781893cd2b296bcfa023a74b962) done

**Review by hryjosn — APPROVED** [review]

**Pull request merged**